packages/timeline-state-resolver/src/service/ConnectionManager.ts

```diff
@@ -0,0 +1,395 @@
+		if (this._updating) return
+		this._updating = true
+
+		const operations: Operation[] = []
+
+		for (const [deviceId, config] of this._config.entries()) {
+			// find connection
+			const connection = this._connections.get(deviceId)
+
+			if (connection) {
+				// see if it should be restarted because of an update
+				if (configHasChanged(connection, config)) {
+					operations.push({ operation: 'update', id: deviceId })
+				} else if (
+					connection.deviceOptions.debug !== config.debug ||
+					connection.deviceOptions.debugState !== config.debugState
+				) {
+					// see if we should set the debug params
+					operations.push({ operation: 'setDebug', id: deviceId })
+				}
+			} else {
+				// create
+				operations.push({ operation: 'create', id: deviceId })
+			}
+		}
+
+		for (const deviceId of this._connections.keys()) {
+			// find if still in config
+			const config = this._config.get(deviceId)
+
+			if (!config) {
+				// not found, so it should be closed
+				operations.push({ operation: 'delete', id: deviceId })
+			}
+		}
+
+		console.log('Ran update, got ops', operations)
+
+		if (operations.length === 0) {
+			// no operations needed
+			this._updating = false
+			return
+		}
+
+		Promise.allSettled(operations.map((op) => this.executeOperation(op))).then(() => {
+			this._updating = false
+			// todo - do another run to verify we have achieved creation of all devices (but prevent us from running into a death loop)
+		})
+	}
+
+	private async executeOperation({ operation, id }: Operation): Promise<void> {
+		// todo: timeout here or log errors?
+
+		switch (operation) {
+			case 'create':
+				await this.createConnection(id)
+				break
+			case 'delete':
+				await this.deleteConnection(id)
+				break
+			case 'update':
+				await this.deleteConnection(id)
+				await this.createConnection(id)
+				break
+			case 'setDebug':
+				await this.setDebugForConnection(id)
+				break
+		}
+	}
+
+	private async createConnection(id: string): Promise<void> {
+		const deviceOptions = this._config.get(id)
+		if (!deviceOptions) return // unexpected - throw error?
+
+		const threadedClassOptions: ThreadedClassConfig = {
+			threadUsage: deviceOptions.threadUsage || 1,
+			autoRestart: false,
+			disableMultithreading: !deviceOptions.isMultiThreaded,
+			instanceName: id,
+			freezeLimit: FREEZE_LIMIT,
+		}
+
+		const container = await createContainer(deviceOptions, id, () => Date.now(), threadedClassOptions) // time out if this gets el stucko
+
+		if (!container) return // todo - log or throw error?
+
+		// set up event handlers
+		this._setupDeviceListeners(id, container)
+
+		this._connections.set(id, container)
+		this.emit('connectionAdded', id, container)
+
+		// trigger device init
+		this._handleConnectionInitialisation(id, container).catch(() => {
+			this.emit('error', 'Device ' + id + ' failed to initialise')
+			this._connections.delete(id) // todo - this will cause device to be recreated next, which may cause a spiral?
```

Comment:
Is more cleanup needed?
This is discarding the thread, but the thread is probably still running at this point?